### PR TITLE
Completed Webhook Class and Added WebhookClient

### PR DESCRIPTION
**Pull request by PurgePJ**
Added as many functionality as possible.

### CODE REVIEW

File: init.lua [hunk 1]
```diff
@@ -15,4 +15,5 @@ return {
 	Stopwatch = require('utils/Stopwatch'),
 	Time = require('utils/Time'),
 	storage = {},
+	WebhookClient = require('client/WebhookClient')
```
Comment (SinisterRectus) [inline init.lua]: Should be alphabetical order after client/Client

File: libs/client/API.lua [hunk 2]
```diff
@@ -653,37 +653,37 @@ function API:getWebhookWithToken(webhook_id, webhook_token) -- not exposed, need
...
+function API:deleteWebhookWithToken(webhook_id, webhook_token) -- (WebhookClient/Webhook):deleteWebhookWithToken
 	local endpoint = f(endpoints.WEBHOOK_TOKEN, webhook_id, webhook_token)
 	return self:request("DELETE", endpoint)
 end
 
-function API:executeWebhook(webhook_id, webhook_token, payload) -- not exposed, needs webhook client
+function API:executeWebhook(webhook_id, webhook_token, payload, file) -- (WebhookClient/Webhook):send
 	local endpoint = f(endpoints.WEBHOOK_TOKEN, webhook_id, webhook_token)
-	return self:request("POST", endpoint, payload)
+	return self:request("POST", endpoint .. "?wait=true", payload, nil, file)
```
Comment (SinisterRectus) [inline libs/client/API.lua]: Waiting should be optional

File: libs/client/Client.lua [hunk 3]
```diff
@@ -122,6 +122,8 @@ local function run(self, token)
 	if not user then
 		return self:error('Could not authenticate, check token: ' .. err1)
 	end
+
+	_G.API = api;
```
Comment (SinisterRectus) [inline libs/client/Client.lua]: Discordia should not use globals.

File: libs/containers/Message.lua [hunk 4]
```diff
@@ -13,7 +14,7 @@ local Message, get = require('class')('Message', Snowflake)
 
 function Message:__init(data, parent)
 	Snowflake.__init(self, data, parent)
-	self._author = self.client._users:_insert(data.author)
+	self._author = self.client._users and self.client._users:_insert(data.author) or data.author
```
Comment (SinisterRectus) [inline libs/containers/Message.lua]: Author should be a User object, not a JSON object.

File: libs/containers/Message.lua [hunk 5]
```diff
@@ -353,7 +353,18 @@ function get.author(self)
 end
 
 function get.channel(self)
-	return self._parent
+	if self._parent.id == self._channel_id then
+		return self._parent
+	else
+		local data, err = self._parent._api:getChannel(self._channel_id)
```
Comment (SinisterRectus) [inline libs/containers/Message.lua]: Property getters should never make HTTP requests.

File: libs/containers/Message.lua [hunk 6]
```diff
@@ -353,7 +353,18 @@ function get.author(self)
...
+		if data then
+			return Channel(data, self)
```
Comment (SinisterRectus) [inline libs/containers/Message.lua]: Channel is an abstract class and should never be instantiated directly.

File: package.lua [hunk 7]
```diff
@@ -22,7 +22,7 @@ SOFTWARE.]]
 
 return {
 	name = 'SinisterRectus/discordia',
-	version = '2.2.0',
+	version = '2.3.0',
```
Comment (SinisterRectus) [inline package.lua]: Version does not get bumped until release (lit publishing).

File: libs/containers/Message.lua [hunk 8]
```diff
@@ -377,7 +388,12 @@ function get.attachments(self)
 end
 
 function get.guild(self)
-	return self._parent.guild
+	if self._parent.guild then
+		return self._parent.guild
+	else	
+		self._parent:error("Unknown Guild Object from WebhookClient Instance.")
+		return {}
```
Comment (SinisterRectus) [inline libs/containers/Message.lua]: This is bad.

File: libs/containers/Message.lua [hunk 9]
```diff
@@ -377,7 +388,12 @@ function get.attachments(self)
 end
 
 function get.guild(self)
-	return self._parent.guild
+	if self._parent.guild then
+		return self._parent.guild
+	else	
+		self._parent:error("Unknown Guild Object from WebhookClient Instance.")
```
Comment (SinisterRectus) [inline libs/containers/Message.lua]: The container API should not log errors like this.

File: libs/client/WebhookClient.lua [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+		return options
+	else
+		return defaultOptions
+	end
+end
+
+local function resolveImage(avatar, id)
+	if avatar then
+		if tostring(avatar):match("^(https?:)") then
+			local res, data = request("GET", avatar)
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Discordia should not make anonymous requests for the end-user.

File: libs/client/WebhookClient.lua [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+	local cleanedUser = {}
+	for k, v in pairs(user) do
+		if k:find("_") then
+			cleanedUser[k:gsub("_", "")] = v
+		end
+	end
+	return cleanedUser
+end
+
+local function modify(self, tbl)
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: This should be a member method for consistency,

File: libs/client/WebhookClient.lua [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+
+	self._options = parseOptions(options)
+	self._id = tostring(id)
+	self._token = tostring(token)
+	self._logger = Logger(self._options.logLevel, self._options.dateTime, self._options.logFile)
+	self._api = API
+	self._webhook = ""
+
+	-- Sync our defaultOptions with the webhook data
+	local webhook, err = self._api:getWebhook(self._id) -- we do not use getWebhookWithToken because it does not return the user object
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Can't make bot API calls without a bot token.

File: libs/client/WebhookClient.lua [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+WebhookClient.__description = "Represents a Webhook Client."
+
+function WebhookClient:__init(id, token, options)
+	Emitter.__init(self)
+
+	self._options = parseOptions(options)
+	self._id = tostring(id)
+	self._token = tostring(token)
+	self._logger = Logger(self._options.logLevel, self._options.dateTime, self._options.logFile)
+	self._api = API
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Did you test this code?

File: libs/client/WebhookClient.lua [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+	WebhookClient[name] = function(self, fmt, ...)
+		local msg = self._logger:log(level, fmt, ...)
+		if #self._listeners[name] > 0 then
+			return self:emit(name, msg or format(fmt, ...))
+		end
+	end
+end
+
+-- Getters
+function get.avatarURL(self, size, ext)
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: `size and ext` don't do anything here.

File: libs/client/WebhookClient.lua [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+function get.name(self)
+	return self._options.name
+end
+
+function get.token(self)
+	return self._webhook.token
+end
+
+function get.user(self)
+	return clean(self._webhook.user) -- does not return a full user object because it is not linked with Client
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Bad.

File: libs/client/WebhookClient.lua [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+-- Setters and Modifiers
+function WebhookClient:setAvatar(avatar)
+	self._options.avatarURL = avatar
+	avatar = resolveImage(avatar) or Resolver.base64(avatar) or null
+
+	local bool, err = self:modify({avatar = avatar})
+	if bool then
+		return self
+	else
+		self:error(err)
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Should return the error instead of logging it.

File: libs/client/WebhookClient.lua [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+end
+
+function WebhookClient:setChannelId(channelId)
+	self._options.channelId = channelId
+	local bool, err = self:modify({channel_id = channelId})
+
+	if bool then
+		return self
+	else
+		self:error(err)
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Should return the error instead of logging it.

File: libs/client/WebhookClient.lua [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+end
+
+function WebhookClient:setName(name)
+	self._options.name = name
+	local bool, err = self:modify({name = name})
+
+	if bool then
+		return self
+	else
+		self:error(err)
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Should return the error instead of logging it.

File: libs/client/WebhookClient.lua [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+	end
+end
+
+function WebhookClient:modify(tbl)
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Should not be exposed as part of the API. See other modifiable objects.

File: libs/client/WebhookClient.lua [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+	local data, err = self._api:executeWebhook(self._id, self._token, { -- We insert these values manually so that we don't send extra (irrevelant) information
+		content = content,
+		username = webhookData.name,
+		avatar_url = webhookData.avatarURL,
+		tts = webhookData.tts,
+		embeds = webhookData.embeds
+	}, webhookData.file)
+
+	if data then
+		return Message(data, self), self
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: I don't know if this should return a Message or just true/false, or a custom WebhookMessage object. Is is possible to write a WebhookClient that does not rely on Discordia containers?

File: libs/client/WebhookClient.lua [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+	else
+		return self:error(err)
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Should return the error instead of logging it.

File: libs/client/WebhookClient.lua [hunk 22]
```diff
@@ -0,0 +1,314 @@
...
+
+function get.user(self)
+	return clean(self._webhook.user) -- does not return a full user object because it is not linked with Client
+end
+
+
+-- Setters and Modifiers
+function WebhookClient:setAvatar(avatar)
+	self._options.avatarURL = avatar
+	avatar = resolveImage(avatar) or Resolver.base64(avatar) or null
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Use the resolver, that's why it's there.

File: libs/client/WebhookClient.lua [hunk 23]
```diff
@@ -0,0 +1,314 @@
...
+
+	local bool, err = self:modify({avatar = avatar})
+	if bool then
+		return self
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Should return a bool.

File: libs/client/WebhookClient.lua [hunk 24]
```diff
@@ -0,0 +1,314 @@
...
+		self:error(err)
+	end
+end
+
+function WebhookClient:setChannelId(channelId)
+	self._options.channelId = channelId
+	local bool, err = self:modify({channel_id = channelId})
+
+	if bool then
+		return self
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Should return a bool.

File: libs/client/WebhookClient.lua [hunk 25]
```diff
@@ -0,0 +1,314 @@
...
+
+	local bool, err = self:modify({avatar = avatar})
+	if bool then
+		return self
+	else
+		self:error(err)
+	end
+end
+
+function WebhookClient:setChannelId(channelId)
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Is this mutable?

File: libs/client/WebhookClient.lua [hunk 26]
```diff
@@ -0,0 +1,314 @@
...
+		self:error(err)
+	end
+end
+
+function WebhookClient:setName(name)
+	self._options.name = name
+	local bool, err = self:modify({name = name})
+
+	if bool then
+		return self
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Should return a bool.

File: libs/client/WebhookClient.lua [hunk 27]
```diff
@@ -0,0 +1,314 @@
...
+
+	if data then
+		return Message(data, self), self
+	else
+		return self:error(err)
+	end
+end
+
+function WebhookClient:delete()
+	local data, err = self._api:deleteWebhook(self._id)
```
Comment (SinisterRectus) [inline libs/client/WebhookClient.lua]: Can't make bot API calls without a bot token.

File: libs/client/WebhookClient.lua (hunk 20 above)
Comment (PurgePJ) [reply libs/client/WebhookClient.lua]: We can return a boolean as of now and work on creating a custom WebhookMessage with the limited data Discord returns when waiting for the object creation.  It is possible to write a WebhookClient without relying on Discordia containers. In my rewrite I don't use any of them in the WebhookClient.

File: libs/containers/Webhook.lua [hunk 28]
```diff
@@ -85,4 +95,122 @@ function get.defaultAvatarURL(self)
...
+function Webhook:delete()
+	local data, err = self.client._api:deleteWebhook(self._id)
+	if data then
+		return true
+	else
+		return false, err
+	end
+end
+
+function Webhook:deleteWithToken()
```
Comment (SinisterRectus) [inline libs/containers/Webhook.lua]: This isn't needed since the delete method already exists.

File: libs/client/API.lua [hunk 29]
```diff
@@ -112,12 +112,18 @@ function API:__init(client)
 	self._mutexes = setmetatable({}, mutexMeta)
 end
 
-function API:authenticate(token)
-	self._headers = {
-		{'Authorization', token},
-		{'User-Agent', f('DiscordBot (%s, %s)', package.homepage, package.version)},
-	}
-	return self:getCurrentUser()
+function API:authenticate(token, isWebhookClient)
```
Comment (SinisterRectus) [inline libs/client/API.lua]: This is not necessary. User-Agent can just be set on initialization. I can make this change when necessary.

File: libs/containers/Webhook.lua [hunk 30]
```diff
@@ -1,52 +1,62 @@
...
+local defaultAvatar = enums.defaultAvatar
+local base64 = ssl.base64
+local readFileSync = fs.readFileSync
+local splitPath = pathjoin.splitPath
 
-function Webhook:__init(data, parent)
-	Snowflake.__init(self, data, parent)
-	self._user = data.user and self.client._users:_insert(data.user) -- DNE if getting by token
+local function resolveImage(avatar, id)
+	if avatar then
```
Comment (SinisterRectus) [inline libs/containers/Webhook.lua]: Why only check for avatar and not id?

File: libs/client/API.lua [hunk 31]
```diff
@@ -109,15 +110,20 @@ local API = require('class')('API')
...
 function API:__init(client)
 	self._client = client
+	self._headers = {
+		{'User-Agent', USER_AGENT}
+	}
 	self._mutexes = setmetatable({}, mutexMeta)
 end
 
-function API:authenticate(token)
+function API:authenticate(token, isWebhookClient)
```
Comment (SinisterRectus) [inline libs/client/API.lua]: Authenticating is not necessary for webhooks.

File: libs/client/API.lua [hunk 32]
```diff
@@ -23,6 +23,7 @@ local BOUNDARY3 = BOUNDARY2 .. '--'
 
 local JSON = 'application/json'
 local MULTIPART = f('multipart/form-data;boundary=%s', BOUNDARY1)
+local USER_AGENT = f('DiscordBot (%s, %s)', package.homepage, package.version)
```
Comment (SinisterRectus) [inline libs/client/API.lua]: Pull the new changes instead of reapplying them here.

File: libs/client/API.lua [hunk 33]
```diff
@@ -109,15 +110,20 @@ local API = require('class')('API')
 
 function API:__init(client)
 	self._client = client
+	self._headers = {
```
Comment (SinisterRectus) [inline libs/client/API.lua]: Pull the new changes instead of reapplying them here.

File: libs/containers/Webhook.lua [hunk 34]
```diff
@@ -1,10 +1,37 @@
 local json = require('json')
 local enums = require('enums')
+local request = require("coro-http").request
```
Comment (SinisterRectus) [inline libs/containers/Webhook.lua]: Remove unused line.

File: libs/containers/Webhook.lua [hunk 35]
```diff
@@ -1,10 +1,37 @@
...
+
 local Snowflake = require('containers/abstract/Snowflake')
 local User = require('containers/User')
-local Resolver = require('client/Resolver')
 
+
+local format = string.format
+local insert, remove = table.insert, table.remove
 local defaultAvatar = enums.defaultAvatar
+local base64 = ssl.base64
```
Comment (SinisterRectus) [inline libs/containers/Webhook.lua]: Remove unused line.

File: libs/containers/Webhook.lua [hunk 36]
```diff
@@ -1,10 +1,37 @@
 local json = require('json')
 local enums = require('enums')
+local request = require("coro-http").request
+local ssl = require('openssl')
```
Comment (SinisterRectus) [inline libs/containers/Webhook.lua]: Remove unused line.

File: libs/containers/Webhook.lua [hunk 37]
```diff
@@ -13,40 +40,17 @@ function Webhook:__init(data, parent)
...
-end
-
-function Webhook:delete()
-	local data, err = self.client._api:deleteWebhook(self._id)
-	if data then
-		return true
-	else
-		return false, err
-	end
+function get.id(self)
```
Comment (SinisterRectus) [inline libs/containers/Webhook.lua]: Remove unnecessary code.

File: libs/containers/Webhook.lua [hunk 38]
```diff
@@ -85,4 +89,110 @@ function get.defaultAvatarURL(self)
...
+	if data then
+		self:_load(data)
+		return true
+	else
+		return false, err
+	end
+end
+
+function Webhook:setAvatar(avatar)
+	self._avatar = avatar and Resolver.base64(avatar) or json.null
```
Comment (SinisterRectus) [inline libs/containers/Webhook.lua]: Stop setting properties before you know they are set.

File: libs/containers/Webhook.lua [hunk 39]
```diff
@@ -13,40 +40,17 @@ function Webhook:__init(data, parent)
...
-		return true
-	else
-		return false, err
-	end
-end
-
+-- Getters
 function Webhook:getAvatarURL(size, ext)
-	return User.getAvatarURL(self, size, ext)
+	return self._avatarURL ~= "" and self._avatarURL or User.getAvatarURL(self, size, ext)
```
Comment (SinisterRectus) [inline libs/containers/Webhook.lua]: Why is this still changed?

File: libs/containers/Webhook.lua [hunk 40]
```diff
@@ -85,4 +89,110 @@ function get.defaultAvatarURL(self)
...
+	local data, err = self:_modify({name = name})
+	if data then
+		return true
+	else
+		return nil, err
+	end
+end
+
+-- Rest of functions
+function Webhook:send(content, options) -- return message
```
Comment (SinisterRectus) [inline libs/containers/Webhook.lua]: Why is this an options table instead of a content table? This should also probably not be named "send".

**Comment by SinisterRectus** [comment]
Once again, I'm spending more time reviewing changes than I would spend writing the code myself.

**Comment by PurgePJ** [comment]
View PR#89.

**Pull request closed without merging**